src/Client.ts

```diff
@@ -0,0 +1,110 @@
+import * as dgram from 'dgram';
+import { Address, IMessage, MessageType } from './Interfaces';
+
+export default class Client
+{
+    private _socket: dgram.Socket;
+    private _address: Address = {
+        ip: 'localhost',
+        port: 8000,
+    };
+    private _connected = false;
+
+    constructor(private _id: number, private _username: string, private _show_log = false)
+    {
+        this._setup_client_socket();
+    }
+
+    connect(server?: Address): Promise<Address>
+    {
+        if (server)
+            this._address = server;
+
+        return Promise.resolve().then(() => {
+            if (this._show_log)
+                console.log(`[C:${this._id}] Registering client.`);
+            return this._internal_send({
+                type: MessageType.REGISTRATION,
+            });
+        })
+        .then(() => {
+            if (this._show_log)
+                console.log(`[C:${this._id}] Connected.`);
+            this._connected = true;
+            return this._address;
+        });
```

Comment:
You may refactor this code with `async/await` syntax: [docs here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function)